File: golang/binary_search.go
func search(nums []int, target int) int {
	start, end := 0, len(nums)-1

	for start <= end {
		middle := (start + end) / 2
		if nums[middle] == target {
			return middle
		} else if nums[middle] < target {
			start = middle + 1
		} else if nums[middle] > target {
			end = middle - 1
		}
	}

	return -1
}

File: typescript/binary_search.ts
function search(nums: number[], target: number): number {
	let start = 0
	let end = nums.length - 1

	while (start <= end) {
		let mid = ~~((start + end) / 2) // Math.floor shorthand
		if (nums[mid] == target) return mid
		else if (nums[mid] > target) end = mid - 1
		else if (nums[mid] < target) start = mid + 1
	}

	return -1
}
